sys/riscv/trap.c

```diff
@@ -0,0 +1,301 @@
+#define KL_LOG KL_VM
+#include <sys/context.h>
+#include <sys/cpu.h>
+#include <sys/errno.h>
+#include <sys/interrupt.h>
+#include <sys/klog.h>
+#include <sys/pmap.h>
+#include <sys/sysent.h>
+#include <sys/thread.h>
+#include <sys/vm_map.h>
+#include <riscv/cpufunc.h>
+#include <riscv/mcontext.h>
+
+/* clang-format off */
+static const char *const exceptions[] = {
+  [SCAUSE_INST_MISALIGNED] = "Misaligned instruction",
+  [SCAUSE_INST_ACCESS_FAULT] = "Instruction access fault",
+  [SCAUSE_ILLEGAL_INSTRUCTION] = "Illegal instruction",
+  [SCAUSE_BREAKPOINT] = "Breakpoint",
+  [SCAUSE_LOAD_MISALIGNED] = "Misaligned load",
+  [SCAUSE_LOAD_ACCESS_FAULT] = "Load access fault",
+  [SCAUSE_STORE_MISALIGNED] = "Misaligned store",
+  [SCAUSE_STORE_ACCESS_FAULT] = "Store access fault",
+  [SCAUSE_ECALL_USER] = "User environment call",
+  [SCAUSE_ECALL_SUPERVISOR] = "Supervisor environment call",
+  [SCAUSE_INST_PAGE_FAULT] = "Instruction page fault",
+  [SCAUSE_LOAD_PAGE_FAULT] = "Load page fault",
+  [SCAUSE_STORE_PAGE_FAULT] = "Store page fault",
+};
+/* clang-format on */
+
+__no_profile static inline bool ctx_interrupt(ctx_t *ctx) {
+  return _REG(ctx, CAUSE) & SCAUSE_INTR;
+}
+
+__no_profile static inline unsigned ctx_code(ctx_t *ctx) {
+  return _REG(ctx, CAUSE) & SCAUSE_CODE;
+}
+
+__no_profile static inline unsigned ctx_intr_enabled(ctx_t *ctx) {
+  return _REG(ctx, SR) & SSTATUS_SPIE;
+}
+
+static __noreturn void kernel_oops(ctx_t *ctx) {
+  unsigned code = ctx_code(ctx);
+  void *epc = (void *)_REG(ctx, PC);
+
+  klog("%s at %p!", exceptions[code], epc);
+
+  switch (code) {
+    case SCAUSE_INST_MISALIGNED:
+    case SCAUSE_INST_ACCESS_FAULT:
+    case SCAUSE_LOAD_MISALIGNED:
+    case SCAUSE_LOAD_ACCESS_FAULT:
+    case SCAUSE_STORE_MISALIGNED:
+    case SCAUSE_STORE_ACCESS_FAULT:
+    case SCAUSE_INST_PAGE_FAULT:
+    case SCAUSE_LOAD_PAGE_FAULT:
+    case SCAUSE_STORE_PAGE_FAULT:
+      klog("Caused by reference to %lx!", _REG(ctx, TVAL));
+      break;
+
+    case SCAUSE_ILLEGAL_INSTRUCTION: {
+      uint32_t badinstr = *(uint32_t *)epc;
+      klog("Illegal instruction %08x in kernel mode!", badinstr);
+      break;
+    }
+
+    case SCAUSE_BREAKPOINT:
+      klog("No debbuger in kernel!");
+      break;
+  }
+  klog("HINT: Type 'info line *%p' into gdb to find faulty code line", epc);
+
+  panic("KERNEL PANIC!!!");
+}
+
+static void page_fault_handler(ctx_t *ctx) {
+  thread_t *td = thread_self();
+
+  unsigned code = ctx_code(ctx);
+  void *epc = (void *)_REG(ctx, PC);
+  vaddr_t vaddr = _REG(ctx, TVAL);
+
+  klog("%s at %p, caused by reference to %lx!", exceptions[code], epc, vaddr);
+
+  pmap_t *pmap = pmap_lookup(vaddr);
+  if (!pmap) {
+    klog("No physical map defined for %lx address!", vaddr);
+    goto fault;
+  }
+
+  vm_prot_t access;
+  if (code == SCAUSE_INST_PAGE_FAULT)
+    access = VM_PROT_EXEC;
+  else if (code == SCAUSE_LOAD_PAGE_FAULT)
+    access = VM_PROT_READ;
+  else
+    access = VM_PROT_WRITE;
```

Comment:
I use the following convention for `if` statements with a single instruction inside:
- if there's a comment beside the statement then I include braces,
- otherwise, I don't use braces.
I applied this convention as it seems to me that it's used throughout the majority of the code (e.g. in corresponding trap modules for other architectures).
I will leave this decision to @cahirwpz.

Reply:
If block is one liner, then it's ok to omit braces. However, if there're at least two lines (even if one of them is a comment) then I'd expect to enclose it with braces.